strawberry/mutation.py

```diff
@@ -1,8 +1,123 @@
+from __future__ import annotations
+
 from functools import partial
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Dict,
+    TypeVar,
+)
+from typing_extensions import Annotated, get_args, get_origin
+
+import strawberry
+from strawberry.annotation import StrawberryAnnotation
+from strawberry.arguments import StrawberryArgument
+from strawberry.extensions.field_extension import (
+    AsyncExtensionResolver,
+    FieldExtension,
+    SyncExtensionResolver,
+)
+from strawberry.utils.str_converters import to_camel_case
+
+from .field import StrawberryField, field
+
+if TYPE_CHECKING:
+    from strawberry.types.info import Info
+
+_T = TypeVar("_T")
+
+
+class InputMutationExtension(FieldExtension):
+    def apply(self, field: StrawberryField) -> None:
+        resolver = field.base_resolver
+        assert resolver
+
+        name = field.graphql_name or to_camel_case(resolver.name)
+        name_captalized = name[0].upper() + name[1:]
+        type_dict: Dict[str, Any] = {
+            "__doc__": f"Input data for `{name}` mutation",
+            "__annotations__": {},
+        }
+        annotations = resolver.wrapped_func.__annotations__
+
+        for arg in resolver.arguments:
+            # Preserve directives
+            annotation = annotations[arg.python_name]
+            if get_origin(annotation) is Annotated:
+                directives = tuple(
+                    d
+                    for d in get_args(annotation)[1:]
+                    if hasattr(d, "__strawberry_directive__")
+                )
+            else:
+                directives = ()
+
```

Comment:
We should think about providing helper methods to make the api more accessible. That's not the scope of this PR though, so lets keep this as a reminder and todo for later. ☺️

Reply:
I actually didn't need it anymore. This code was ported from the relay PR, which was ported from my strawberry-django-plus lib. At that time strawberry probably didn't have `StrawberryArgument.directives`, but now it has :)